Sources/MapboxMaps/Location/LocationInterpolator.swift

```diff
@@ -0,0 +1,72 @@
+internal protocol LocationInterpolatorProtocol: AnyObject {
+    func interpolate(from fromLocation: InterpolatedLocation,
+                     to toLocation: InterpolatedLocation,
+                     percent: Double) -> InterpolatedLocation
+}
+
+internal final class LocationInterpolator: LocationInterpolatorProtocol {
+
+    private let interpolator: InterpolatorProtocol
+    private let directionInterpolator: InterpolatorProtocol
+    private let latitudeInterpolator: InterpolatorProtocol
+
+    internal init(interpolator: InterpolatorProtocol,
+                  directionInterpolator: InterpolatorProtocol,
+                  latitudeInterpolator: InterpolatorProtocol) {
+        self.interpolator = interpolator
+        self.directionInterpolator = directionInterpolator
+        self.latitudeInterpolator = latitudeInterpolator
+    }
+
+    internal func interpolate(from fromLocation: InterpolatedLocation,
+                              to toLocation: InterpolatedLocation,
+                              percent: Double) -> InterpolatedLocation {
+        let course: CLLocationDirection?
+        if let startCourse = fromLocation.course,
+           let endCourse = toLocation.course {
+            course = directionInterpolator.interpolate(
+                from: startCourse,
+                to: endCourse,
+                percent: percent)
+        } else if let endCourse = toLocation.course {
+            course = endCourse
+        } else {
+            course = nil
+        }
+
+        let heading: CLLocationDirection?
+        if let startHeading = fromLocation.heading,
+           let endHeading = toLocation.heading {
+            heading = directionInterpolator.interpolate(
+                from: startHeading,
+                to: endHeading,
+                percent: percent)
+        } else if let endHeading = toLocation.heading {
+            heading = endHeading
+        } else {
+            heading = nil
+        }
```

Comment:
do you interpolate both the heading and course at the same time, even if one is not used?
And I assume this LocationInterpolator is shared between viewport and location manager, and by interpolating both, you can achieve setting the puck bearing source separately on viewport and location..

Reply:
> do you interpolate both the heading and course at the same time, even if one is not used?
Yes
> And I assume this LocationInterpolator is shared between viewport and location manager, and by interpolating both, you can achieve setting the puck bearing source separately on viewport and location..
That's correct.